### PR TITLE
ref(notifications): Reorganize the notifications module

### PR DESCRIPTION
**Pull request by mgaeta**
Reorganizing the notifications hierarchy so that there is less repeated code and it is easy to add the Digest Notification.

### CODE REVIEW

**Comment by ceorourke** [comment]
Could you include a before and after of the file structure? I'm not sure I understand why it's now `/notifications/notifications/*`. There also seems to be some refactoring in this PR which I'm guessing is because it was previously part of another one.

**Comment by mgaeta** [comment]
Before:

```
sentry/notifications/
  activity/
    assigned.py
    base.py
    ...
    unassigned.py
  utils/
    avatar.py
    ...
    participants.py
  base.py
  digest.py
  helpers.py
  manager.py
  notify.py
  rules.py
  types.py
  user_report.py
```

After:
```
sentry/notifications/
  notifications/
    activity/
      assigned.py
      base.py
      ...
      unassigned.py
    base.py
    digest.py
    rules.py
    user_report.py
  utils/
    avatar.py
    ...
    participants.py
  helpers.py
  manager.py
  notify.py
  types.py
```

So everything in the `BaseNotification` hierarchy is in the `sentry/notifications/notifications/` directory.

File: src/sentry/mail/notifications.py [hunk 1]
```diff
@@ -21,48 +21,48 @@ def get_headers(notification: BaseNotification) -> Mapping[str, Any]:
...
+    context: Optional[Mapping[str, Any]] = None,
+    mail_option_key: Optional[str] = None,
 ) -> bytes:
     key = mail_option_key or "mail:subject_prefix"
     prefix = str(
         ProjectOption.objects.get_value(notification.project, key)
         or options.get("mail.subject-prefix")
     )
-    return f"{prefix}{notification.get_subject()}".encode("utf-8")
+    return f"{prefix}{notification.get_subject(context)}".encode("utf-8")
```
Comment (ceorourke) [inline src/sentry/mail/notifications.py:47]: was this (stuff about context) supposed to be in another PR?

File: src/sentry/mail/notifications.py [hunk 2]
```diff
@@ -21,48 +21,48 @@ def get_headers(notification: BaseNotification) -> Mapping[str, Any]:
...
-
-
 def log_message(notification: BaseNotification, user: User) -> None:
     extra = {
         "project_id": notification.project.id,
         "user_id": user.id,
     }
-    if notification.group:
-        extra.update({"group": notification.group.id})
+    group = getattr(notification, "group", None)
```
Comment (ceorourke) [inline src/sentry/mail/notifications.py:63]: this also seems like maybe it was from the typing PR?

Comment (mgaeta) [reply src/sentry/mail/notifications.py:63]: I reorganized the classes so that only some of them have the `group` attribute. For example, the deploy notification.

File: src/sentry/mail/notifications.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/mail/notifications.py:47]: I think it makes sense to include here but I can extract it to the digest notification PR.

File: src/sentry/notifications/notifications/activity/new_processing_issues.py [hunk 3]
```diff
@@ -25,19 +25,25 @@ def get_participants_with_group_subscription_reason(
             for provider, users in users_by_provider.items()
         }
 
+    def get_activity_name(self) -> str:
+        # TODO MARCOS make sure this doesn't cause an error
```
Comment (ceorourke) [inline src/sentry/notifications/notifications/activity/new_processing_issues.py:29]: did you mean to leave this in?

File: src/sentry/notifications/notifications/user_report.py [hunk 4]
```diff
@@ -38,12 +38,16 @@ def get_category(self) -> str:
...
-    def get_subject(self) -> str:
+    def get_subject(self, context: Optional[Mapping[str, Any]] = None) -> str:
         # Explicitly typing to satisfy mypy.
         message = f"{self.group.qualified_short_id} - New Feedback from {self.report['name']}"
         message = force_text(message)
         return message
 
+    def get_notification_title(self) -> str:
+        # This shouldn't be possible but adding a message just in case.
+        return "Weekly Report"
```
Comment (ceorourke) [inline src/sentry/notifications/notifications/user_report.py:49]: this isn't the weekly report, it's user feedback

**Review by ceorourke — APPROVED** [review]
